docs/architecture/architecture.md

```diff
@@ -0,0 +1,671 @@
+forward the process exit code to the `shim`. The `shim` will then disconnect
+from `cc-proxy` and terminates as well.
+
+One corner case `cc-runtime` also needs to handle is when it receives a signal
+for a container that's not yet running. In that case, `cc-runtime` simply stops
+`cc-shim` for the `SIGTERM` or `SIGKILL` signals and does nothing for all other
+signals.
+
+#### [`delete`](https://github.com/clearcontainers/runtime/blob/master/delete.go)
+
+`delete` is about deleting all resources held by a stopped/killed container.
+Running containers can not be deleted unless the OCI runtime is explictly being
+asked to. In that case it will first `kill` the container and only then `delete`
+it.
+
+The `delete` code path differs significantly between having to delete one container
+inside a pod and having to delete an entire pod. In the former case, `cc-runtime`
+will only send a `SIGKILL` signal to the container process while the latter case is
+about completely deleting the whole thing: the pod, its containers, all `cc-shim`
+instances and the virtual machine itself.
+Also, the latter case is only supported by Kubernetes whereas Docker always asks
+for the former one.
+
+Below we will focus on describing the entire pod deletion process and we will
+look at the generic, multi-containers pod setup. Docker is just a particular case
+for single container pods.
+
+1. `cc-runtime` connects to `cc-proxy` and sends it the `attach` command to let
+  it know on which pod the container it is trying to to `delete` is running.
+2. `cc-runtime` terminates all container processes within a pod by sending the
+  `KILLCONTAINER` command to all of them.
+3. As each container process termination will be forwarded to all associated `cc-shim`
+  instances, `cc-runtime` waits for all of them to terminate as well.
+4. `cc-runtime` sends the `UnregisterVM` command to `cc-proxy`, to let it know
+  that the virtual machine that used to host the pod should no longer be used.
+5. `cc-runtime` explicitly shuts the virtual machine down.
+6. The host namespaces are cleaned up and destroyed. In particular, `cc-runtime`
+  offloads the networking namespace cleanup path by calling into the specific
+  networking model (CNM or CNI) removal method.
+7. All remaining pod related resources on the host are deleted.
+
+## Proxy
+
+`cc-proxy` is a daemon offering access to the VM [`cc-agent`](https://github.com/clearcontainers/agent)
+to multiple `cc-shim` and `cc-runtime` clients. Only a single instance of `cc-proxy`
+per host is necessary as it can be used for several different VMs.
+Its main role is to:
+- Arbitrate access to the `cc-agent` control channel between all the `cc-runtime`
+  instances and the `cc-shim` ones.
+- Route the I/O streams and signals between the various `cc-shim` instances and
+  the `cc-agent`.
+
+`cc-proxy` API is available through a single socket for all `cc-shim` and `cc-runtime`
```

Comment:
Insert a 'The' at the start of the sentence? - 'The `cc-proxy` API...'

Reply:
ack